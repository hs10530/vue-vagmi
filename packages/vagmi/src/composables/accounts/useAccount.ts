import type { GetAccountResult } from '@wagmi/core'
import { getAccount, watchAccount } from '@wagmi/core'
import { useQuery, useQueryClient } from 'vue-query'
import type { UseQueryOptions } from 'vue-query'
import { tryOnScopeDispose } from '@vueuse/core'

export type UseAccountConfig = Pick<
  UseQueryOptions<GetAccountResult, Error>,
  'suspense' | 'onError' | 'onSettled' | 'onSuccess'
>

export const queryKey = () => [{ entity: 'account' }] as const

const queryFn = () => {
  const result = getAccount()
  if (result.address)
    return result
  return null
}

export function useAccount({
  suspense,
  onError,
  onSettled,
  onSuccess,
}: UseAccountConfig = {}) {
  const queryClient = useQueryClient()

  const accountQuery = useQuery(queryKey(), queryFn, {
    staleTime: 0,
    suspense,
    onError,
    onSettled,
    onSuccess,
  })

  const unwatch = watchAccount((data) => {
    // TODO: Fix reactivity
    queryClient.setQueryData(queryKey(), data?.address
      ? data
      : {
          address: null,
          connector: null,
        })
    accountQuery.refetch()
  })

  tryOnScopeDispose(() => {
    unwatch()
  })

  return accountQuery
}
