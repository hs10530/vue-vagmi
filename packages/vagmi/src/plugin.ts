import type {
  ClientConfig,
  Provider,
  Client as VanillaClient,
  WebSocketProvider,
} from '@wagmi/core';
import {
  createClient as createVanillaClient,
} from '@wagmi/core';
import type {
  InjectionKey,
  Plugin,
  Ref,
} from 'vue';
import {
  inject,
  readonly,
  shallowRef,
  triggerRef,
} from 'vue';
import { QueryClient, VueQueryPlugin } from 'vue-query';

export type CreateClientConfig<
  TProvider extends Provider = Provider,
  TWebSocketProvider extends WebSocketProvider = WebSocketProvider,
> = ClientConfig<TProvider, TWebSocketProvider> & {
  queryClient?: QueryClient
};

export function createClient<
  TProvider extends Provider,
  TWebSocketProvider extends WebSocketProvider,
>({
  queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        cacheTime: 1_000 * 60 * 60 * 24, // 24 hours
        // TODO: uncomment when persistor becomes available
        networkMode: 'offlineFirst',
        refetchOnWindowFocus: false,
        retry: 0,
      },
      mutations: {
        // TODO: uncomment when persistor becomes available
        networkMode: 'offlineFirst',
      },
    },
  }),
  ...config
}: CreateClientConfig<TProvider, TWebSocketProvider> = {}) {
  const client = createVanillaClient<TProvider, TWebSocketProvider>(config);
  // TODO: Add persistor when it becomes available
  return Object.assign(client, { queryClient });
}

export type Client<
  TProvider extends Provider = Provider,
  TWebSocketProvider extends WebSocketProvider = WebSocketProvider,
> = VanillaClient<TProvider, TWebSocketProvider> & { queryClient: QueryClient };

export const VagmiClientKey: InjectionKey<Ref<Client>> = Symbol('vagmi');

export function VagmiPlugin(client = createClient()): Plugin {
  return {
    install(app) {
      // Setup vue-query
      app.use(VueQueryPlugin, {
        queryClient: client.queryClient,
      });

      const reactiveClient = shallowRef(client);

      // Setup @wagmi/core
      if (client.config.autoConnect)
        client.autoConnect();

      const unsubscribe = client.subscribe(() => {
        triggerRef(reactiveClient);
      });

      const originalUnmount = app.unmount;
      app.unmount = function vagmiUnmount() {
        unsubscribe();
        originalUnmount();
      };

      app.provide(VagmiClientKey, reactiveClient);
    },
  };
}

export function useClient() {
  const vagmiClient = inject(VagmiClientKey);
  if (!vagmiClient) {
    throw new Error(
      [
        '`useClient` must be used within `WagmiConfig`.\n',
        'Read more: https://wagmi.sh/docs/WagmiConfig',
      ].join('\n'),
    );
  }

  return vagmiClient;
}

export function useReadonlyClient(): Readonly<ReturnType<typeof useClient>> {
  const client = useClient();
  return readonly(client) as any;
}
