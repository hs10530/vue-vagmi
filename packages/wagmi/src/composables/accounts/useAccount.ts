import type { GetAccountResult } from '@wagmi/core'
import { getAccount, watchAccount } from '@wagmi/core'
import { useQuery, useQueryClient } from 'vue-query'
import type { UseQueryOptions } from 'vue-query'
import { onScopeDispose, reactive } from 'vue'

export type UseAccountConfig = Pick<
  UseQueryOptions<GetAccountResult, Error>,
  'suspense' | 'onError' | 'onSettled' | 'onSuccess'
>

export const queryKey = () => [{ entity: 'account' }] as const

const queryFn = () => {
  const result = getAccount()
  if (result.address)
    return result
  return null
}

export function useAccount({
  suspense,
  onError,
  onSettled,
  onSuccess,
}: UseAccountConfig = {}) {
  const queryClient = useQueryClient()

  const options = reactive({
    queryKey: queryKey(),
    queryFn,
    suspense,
    onError,
    onSettled,
    onSuccess,
  })

  const accountQuery = useQuery(options)

  const unwatch = watchAccount((data) => {
    queryClient.setQueryData(queryKey(), data?.address ? data : null)
  })

  onScopeDispose(unwatch)

  return accountQuery
}
