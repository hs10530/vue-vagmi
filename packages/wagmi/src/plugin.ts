import type { providers } from 'ethers'
import type {
  WagmiClient,
  ClientConfig as WagmiClientConfig,
} from '@wagmi/core'
import {
  createClient as createWagmiClient,
} from '@wagmi/core'
import type {
  InjectionKey,
  Plugin,
} from 'vue'
import {
  inject,
} from 'vue'
import { QueryClient, VueQueryPlugin } from 'vue-query'

export type DecoratedWagmiClient<
  TProvider extends providers.BaseProvider = providers.BaseProvider,
  TWebSocketProvider extends providers.WebSocketProvider = providers.WebSocketProvider,
> = WagmiClient<TProvider, TWebSocketProvider> & { queryClient: QueryClient }

export type ClientConfig<
  TProvider extends providers.BaseProvider = providers.BaseProvider,
  TWebSocketProvider extends providers.WebSocketProvider = providers.WebSocketProvider,
> = WagmiClientConfig<TProvider, TWebSocketProvider> & {
  queryClient?: QueryClient
}

export function createClient<
  TProvider extends providers.BaseProvider,
  TWebSocketProvider extends providers.WebSocketProvider,
>({
  queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        cacheTime: 1_000 * 60 * 60 * 24, // 24 hours
        // TODO: uncomment when persistor becomes available
        // networkMode: 'offlineFirst',
        refetchOnWindowFocus: false,
        retry: 0,
      },
      mutations: {
        // TODO: uncomment when persistor becomes available
        // networkMode: 'offlineFirst',
      },
    },
  }),
  ...config
}: ClientConfig<TProvider, TWebSocketProvider> = {}) {
  const client = createWagmiClient<TProvider, TWebSocketProvider>(config)
  // TODO: Add persistor when it becomes available
  return Object.assign(client, { queryClient })
}

export const VagmiClientKey: InjectionKey<DecoratedWagmiClient> = Symbol('wagmi')

export function updateState(
  state: Record<string, unknown>,
  update: Record<string, any>,
): void {
  Object.keys(state).forEach((key) => {
    state[key] = update[key]
  })
}

export function VagmiPlugin(client = createClient()): Plugin {
  return {
    install(app) {
      // Setup vue-query
      app.use(VueQueryPlugin, {
        queryClient: client.queryClient,
      })

      // Setup @wagmi/core
      if (client.config.autoConnect)
        client.autoConnect()

      app.provide(VagmiClientKey, client)
    },
  }
}

export function useClient() {
  const wagmiClient = inject(VagmiClientKey)
  if (!wagmiClient)
    throw new Error('Must be used within VueWagmiPlugin')

  return wagmiClient
}
