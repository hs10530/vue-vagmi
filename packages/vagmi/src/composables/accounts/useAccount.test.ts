import { connect } from '@wagmi/core'
import { isRef } from 'vue'
import { renderComposable, setupClient } from '../../../test'
import { VagmiPlugin } from '../../plugin'
import { useAccount } from './useAccount'
import type { UseQueryReturnType } from '../utils/useQuery'

function unrefAllProperties<T>(result: UseQueryReturnType<T, Error>) {
  const realValues = {}
  Object.keys(result).forEach((key) => {
    // @ts-expect-error: Internal
    realValues[key] = isRef(result[key]) ? result[key].value : result[key]
  })
  return realValues as UseQueryReturnType<T, Error>
}

describe('useAccount', () => {
  describe('mounts', () => {
    it('is connected', async () => {
      const client = setupClient()
      await connect({ connector: client.connectors[0] })

      const { result, waitFor } = renderComposable(() => useAccount(), {
        global: {
          plugins: [
            VagmiPlugin(client)
          ]
        }
      })

      await waitFor(() => result.isSuccess.value)

      const { internal, ...rest } = result

      expect(unrefAllProperties(rest)).toMatchSnapshot()
    })

    it('is not connected', async () => {
      const client = setupClient()

      const { result, waitFor } = renderComposable(() => useAccount(), {
        global: {
          plugins: [
            VagmiPlugin(client)
          ]
        }
      })

      await waitFor(() => result.isSuccess.value)

      const { internal, ...rest } = result

      expect(unrefAllProperties(rest)).toMatchSnapshot()
    })
  })
})
