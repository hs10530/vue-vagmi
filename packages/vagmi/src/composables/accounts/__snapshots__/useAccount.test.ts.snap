// Vitest Snapshot v1

exports[`useAccount > mounts > is connected 1`] = `
{
  "data": {
    "address": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    "connector": "<MockConnector>",
  },
  "error": null,
  "errorUpdateCount": 0,
  "fetchStatus": "idle",
  "isError": false,
  "isFetched": true,
  "isFetching": false,
  "isLoading": false,
  "isRefetching": false,
  "isSuccess": true,
  "refetch": [Function],
  "status": "success",
  "suspense": [Function],
}
`;

exports[`useAccount > mounts > is not connected 1`] = `
{
  "data": null,
  "error": null,
  "errorUpdateCount": 0,
  "fetchStatus": "idle",
  "isError": false,
  "isFetched": true,
  "isFetching": false,
  "isLoading": false,
  "isRefetching": false,
  "isSuccess": true,
  "refetch": [Function],
  "status": "success",
  "suspense": [Function],
}
`;
